Sc update

MRossol opened the pull request:
Was actually pretty straightforward. Only had to make two changes:
1) If trans_cap_cost in trans_table don't compute costs
2) Recursion magic to merge sc_points and trans_tables by capacity!

Also some general clean-up.

Still need to test/write tests, waiting on data from @mikebannis

MRossol commented:
closes #311

grantbuster commented on reV/supply_curve/supply_curve.py:
```diff
@@ -256,30 +269,51 @@ def _merge_sc_trans_tables(cls, sc_points, trans_table,
+                        compare_cap=False,
+                        sc_cols=sc_cols))
+
+                trans_sc_table = pd.concat(trans_sc_table)
```
regarding syntax and read-ability, I tend to think this spacing method isn't appropriate here. I understand that you want an empty space after for-loops, but having the pd.concat() call directly above the else: statement visually joins the two, while in fact it would be more intuitive to see the pd.concat() grouped with the for-loop instead of the else: indent level. Does that make sense? If you like it this way don't feel obligated to change it, but I just dont believe in "absolute" rules in cases like this.

MRossol replied:
I think thats fair and a good "exception"

mikebannis replied:
Is there an easy way to reduce the total amount of indentation? I'm looking at this with a lot of github comments in it, but it's hard to see what lines up with what.

grantbuster replied:
I would almost always suggest having the code side by side with the github markup. The github sequential markup (instead of side by side) is really difficult to read. 

As far as reducing indentation: write more functions!

MRossol replied:
And ruin my recursion!?!?!

grantbuster replied:
haha no no, just general comment about lots of indentation

grantbuster commented on reV/supply_curve/supply_curve.py:
```diff
@@ -256,30 +269,51 @@ def _merge_sc_trans_tables(cls, sc_points, trans_table,
+                        sc_points.loc[mask].copy(),
+                        table,
+                        compare_cap=False,
+                        sc_cols=sc_cols))
```
I think i understand what this is doing. Is it okay if there are duplicates afterward? Is there any check for duplicates if not?

MRossol replied:
Good catch, this is what I realized when we were chatting earlier. I've had Bannister add in a min_cap column and updated the code to make the bins non-overlapping

MRossol commented on reV/handlers/transmission.py:
```diff
@@ -51,7 +51,7 @@ def __init__(self, trans_table, line_tie_in_cost=14000, line_cost=3667,
         logger.debug('Trans table input: {}'.format(trans_table))
         logger.debug('Line tie in cost: {} $/MW'.format(line_tie_in_cost))
-        logger.debug('Line cost: {} $/MW-mile'.format(line_cost))
+        logger.debug('Line cost: {} $/MW-km'.format(line_cost))
```
Yup! At least thats the hope plan...

MRossol commented on reV/handlers/transmission.py:
```diff
@@ -148,6 +148,14 @@ def _parse_table(trans_table):
 
+        trans_table = \
+            trans_table.rename(columns={'trans_line_gid': 'trans_gid',
+                                        'trans_gids': 'trans_line_gids'})
```
So originally the naming was as follows which IMO made no sense:
`trans_line_gid` == the transmission feature id, No idea why "line" was in the name
I've had Bannister re-name this column to `trans_gid` in the new tables

`trans_gids` == the list of transmission line gids that connect to the feature (Substations only). No idea why line wasn't in this name so I added it in. Also wanted to make it less ambiguous with respect to `trans_gid`

grantbuster replied:
Gotcha and totally agree. Can you add an in-line comment describing the feature gid vs. line-gids (for substations)? This is going to appear very confusing as soon as next week.

MRossol replied:
# Update legacy transmission table columns to match new less ambiguous
        # column names:
        # trans_gid -> the transmission feature id, legacy name: trans_line_gid
        # trans_line_gids -> gids of transmission lines connected to the given
        # transmission feature (only used for Substations),
        # legacy name: trans_gids

MRossol commented on reV/supply_curve/supply_curve.py:
```diff
@@ -49,7 +49,7 @@ class SupplyCurve:
         transmission_multiplier and the tie-in cost. Default units are $/MW.
-    dist_mi : float
+    dist_km : float
         Distance in miles from supply curve point to transmission connection.
```
dop! Will do

MRossol commented on reV/supply_curve/supply_curve.py:
```diff
@@ -68,9 +68,10 @@ def __init__(self, sc_points, trans_table, fcr, sc_features=None,
             Path to .csv or .json or DataFrame containing supply curve
-            transmission mapping
+            transmission mapping, can also be a list of transmission tables
+            with different line voltage (capacity) ratings.
```
Good call I'll add that in

MRossol commented on reV/supply_curve/supply_curve.py:
```diff
@@ -244,7 +253,11 @@ def _merge_sc_trans_tables(cls, sc_points, trans_table,
+             line voltage (capacity) or pre-loaded dataframe).
+        compare_cap : bool, optional
+            Flag to compare the sc_point and transmission feature capacities,
+            'capacity' <= 'max_cap'
```
So this came up during the recursion. Originally I was going to drop the min/max_cap columns from the trans_tables but figured it would be good to keep those for debugging. The flag isn't exposed to the user its only used inside the method during the recursion. Its on by default but turned off here:  https://github.com/NREL/reV/blob/c42b431184d6f8ce9ea3bcb50472d9cbe45d3955/reV/supply_curve/supply_curve.py#L289

when we are iterating through the different capacity "bins" This is to make sure we don't end up in a recursion loop

grantbuster replied:
hmmm okay. the reason for this still kind of confuses me. the goal here is to join transmission tables with SC agg tables for different max capacities? Could we also do a left join and drop duplicates based on the capacity comparison? That would reduce one confusing level of recursion and also make a natural location to check for duplicates.

MRossol replied:
I don't follow how that would work. The inputs are as follows:
sc_gids of interest
each trans_table contains all sc_points and their connection to the transmission features in the given bin (> min_cap, <= max_cap)

We want to pull the transmission feature from the proper table for each sc_gid based on the sc_gids capacity. I don't follow how we can properly delete duplicates. Do we just join the tables and then delete duplicate trans_gids based on the transmission capacity (retaining the smallest capacity feature?)

MRossol commented on reV/supply_curve/supply_curve.py:
```diff
@@ -256,30 +269,51 @@ def _merge_sc_trans_tables(cls, sc_points, trans_table,
+        if isinstance(trans_table, (list, tuple)):
+            trans_sc_table = []
+            for table in trans_table:
+                trans_sc_table.append(cls._merge_sc_trans_tables(
```
See earlier comment

MRossol commented:
@grantbuster I think this should be ready for Travis. I added tests to check against a baseline and to check to make sure that the final sc table is connecting sc_points to features in the correct capacity bins.

One NOTE for Travis, I updated the trans_cap_cost column to be trans_cap_cost_mw. This is to help distinguish the costs coming in from Mike and the costs computed in reV based off of the sc_gid capacity.

grantbuster commented:
Okay, looks like tests are still failing so lets let travis run off of a dev branch. 

is `trans_cap_cost_mw` a dollars per MW unit? I would suggest including the `per_mw`, as a numerator/denominator confusion could be disastrous.

MRossol commented:
I'll fix the tests now, but good call. I'll change it to _per_mw

grantbuster commented on reV/supply_curve/supply_curve.py:
```diff
@@ -855,12 +855,6 @@ def full_sort(self, sort_on='total_lcoe',
-        if 'offshore' in supply_curve:
-            sum_cols = {'combined_cap_cost': ['array_cable_CAPEX',
-                                              'export_cable_CAPEX',
-                                              'trans_cap_cost_per_mw']}
```
@MRossol fyi i think we talked about this previously but i remember what this was for. It was to sum up some of the cabling costs to get an idea of the offshore transmission capex. You can do most of this in NRWAL now and manually sum some columns in the SC table otherwise. I think it's safe to remove and the hard-coded column names were causing travis problems.

MRossol replied:
Sounds good to me and glad Travis' bug wasn't "me"

grantbuster replied:
I wasnt blaming you but i think its good to remove the hard-coded col names nonetheless

MRossol replied:
I wasn't saying you were, I was just worried i'd fucked up seeing as i pushed that at 10pm last night and the first thing I saw when logging on was, "it failed"...

grantbuster replied:
oh haha yeah gotcha, might have misread. yeah the tests are still failing but thats unrelated i think.

grantbuster reviewed (commented):
What was the cause of the discrepancy on this?

MRossol commented:
One the baseline tests? I'm 99% sure the difference was in the file ordering when multiple tables are loaded. I noticed that there are a few transmission features with identical cost and the new tests were connecting to a different feature. Seems like it was a strange sort change.
